### PR TITLE
feat: create mqtt client

### PR DESCRIPTION
**Pull request by Alxandr**
BREAKING-CHANGE: rename most of the crates

Remaining work:
- [x] `Into<Topic>` for `&CommandTopic` and `&StateTopic`.
- [x] Allow explicit control over MQTT topics.
- [x] Fix broken doc links

### CODE REVIEW

**Comment by Alxandr** [comment]
@arcnmx do you care about (temporarily) dropping Windows support?

**Comment by arcnmx** [comment]
Heh, personally my use-case [already requires dbus and systemd](https://github.com/arcnmx/systemd2mqtt), so I’m not exactly affected!

However, I do have a few things to note about paho-mqtt:

1.  Without setting `default-features = false` it won’t be possible to [turn off the `bundled` feature](https://github.com/eclipse/paho.mqtt.rust/blob/eee12c62aaba6309f95fb15ba1ebe1a2eb721e6e/Cargo.toml#L17), which is bad form for a library crate. I think the solution here is to expose a `paho-bundled = ["paho-mqtt/bundled"]` feature, and include it by default in the mqtt-client crate.
2.  Rust-native alternatives like [rumqttc](https://github.com/bytebeamio/rumqtt) may be worth considering, I’ve been thinking about switching to it recently.
3.  If you really want to go insane with generics, build the whole thing on top of `futures::{Stream,Sink}` and make the mqtt backend pluggable :P

**Comment by Alxandr** [comment]
I think keeping paho as default feature is (at least for now) a good idea - as it's the only way you can use the library (I've - for now - sealed the traits). I looked at `rumqttc` - but they claimed that v5 support wasn't there yet - so I figured I'd keep the door open by going a trait route, but not including it at the moment.

Basically - as it stands, `hass-mqtt-client` is `cfg(unix)` only - the only reason to make `paho` optional is so that CI doesn't fail on windows.

**Comment by Alxandr** [comment]
Welp - I'm back to windows builds failing (shocker)... Turns out that `target.cfg(unix).features` does not work :-/. Guess I'll have to figure out how to allow the windows builds to fail.

**Comment by arcnmx** [comment]
> Turns out that `target.cfg(unix).features` does not work :-/.

This should work just fine:

```toml
[target.'cfg(not(target_env = "msvc"))'.dependencies]
paho-mqtt = { version = "0.11", default-features = false, optional = true }

[features]
default = ["paho", "paho-bundled"]
paho = ["dep:paho-mqtt"]
paho-bundled = ["paho-mqtt?/bundled"]
```

(cfgs in the source will need adjustment to match, since the presence of the `paho` feature doesn't mean the dependency is actually enabled anymore - either include `target_env` in all cfgs referencing `paho` or use a build script to set a common cfg var)

btw, entirely unrelated question, why do you not like `mod.rs`?

**Comment by Alxandr** [comment]
I also have no love for C dependencies 😛. As for MQTT v5 - I don't remember anymore why I thought that was required - so it may not be a big issue. The MQTT traits required for `hass-mqtt-client` will be pretty small - so we can see once that's "done" what features it actually uses.

As for mod.rs, there's a couple of reasons I don't "like" it (I have nothing against it mind you):
1. Consistency - I would either go for always using mod.rs for modules, or never.
2. "Magic" - mod.rs has a "special" meaning that people have to know about. Compared to not using it, you will always find a file with the matching name of the `mod` statement. `lib.rs` suffers from the same problem - but there's still one less at lest.
3. The directory structure you see in this project is what rust-analyzer creates for me. I don't make the files myself. I write `mod foo`, and then press control+. on it, and select "create module" (or whatever the option is called).

As someone who also does a lot of frontend (js/ts) development - I know some of the issues that can arise from all your components living in `index.js` files - so I prefer keeping distinct file-names where I can.

**Comment by Alxandr** [comment]
@arcnmx not sure if you're interested in the hass-mqtt-client or not - but I think the public API is about where I want it for now (though I may have forgotten parts). The implementation however..... is probably to be replaced. It needs to be refactored and cleaned. But I think I want to try building on it a bit first.

Also - the entire topic router was just something I wrote mostly for fun, to see if I could figure something cool out. I'm probably going to replace it in it's entirely with a `HashMap<String, Vec<T>>` - but it was fun to figure out :).

**Comment by Alxandr** [comment]
@arcnmx I've done some more rather major refactorings (mostly just renaming things - including crates). I will probably reshuffle some more in the MQTT types, cause I found that I probably need state and discovery types to live in the same crate. I've also create an example that showcases a fake "light" (not tested it with a real homeassistant yet though, just looked at the MQTT messages it sends). If you're interested in the `HassMqttClient`, please take a look at https://github.com/YoloDev/hass-rs/blob/8fe31151f8125982faf543cf92e03d5bb00c439a/examples/mqtt-light/src/main.rs and let me know what you think :).

**Comment by arcnmx** [comment]
So Iʼve been taking a look at porting to it, hereʼs a few early observations:

1.  Lots of `warning: unresolved link to LightColorState::green` and related errors when generating docs
2.  I had to apply my [`default-features = false` changes](https://github.com/arcnmx/hass-rs/commit/efa0f6176c113f6973dd7531db945cda66ba32ca) from my [earlier comment](https://github.com/YoloDev/hass-rs/pull/26#issuecomment-1332644766) to stop it from using the vendoring/bundled paho build.
3.  `error_stack::Result` makes working with the library awkward, as it seems to force you to adopt it for the entire applicationʼs error handling. This seems to be because `error_stack::Report<C>` doesnʼt impl `std::error::Error`, so it isnʼt compatible with any other error handling libraries.
    It might be a bit inappropriate to use in a library, the same way anyhow is? Iʼll probably just try switching to it for now so I can `?` again.
    - It's also a little insidious in that the `Result<T, E>` return types in the documentation look normal, but then you realise that they're very much not!
5.  The majority of the client crateʼs types donʼt seem to be exported publicly because all the modules are private - generate docs and look at all the broken links. For example: `EntityTopic`, `CommandTopic`, `ConnectError`, `CreateEntityError`, etc. Consider making the modules public?
6.  Small nitpick, but `MqttQosLevel` probably doesnʼt need the `Mqtt` prefix

**Comment by Alxandr** [comment]
> 1. Lots of `warning: unresolved link to LightColorState::green` and related errors when generating docs

Thanks. Didn't notice that. Will take a look and fix it :).

> 2. I had to apply my [`default-features = false` changes](https://github.com/arcnmx/hass-rs/commit/2dbee9d34acab5743392f3c35ef13d92a20cf923) from my [earlier comment](https://github.com/YoloDev/hass-rs/pull/26#issuecomment-1332644766) to stop it from using the vendoring/bundled paho build.

Yeah - the `features` block isn't done. I'll look at the commit you made.

> 3. `error_stack::Result` makes working with the library awkward, as it seems to force you to adopt it for the entire applicationʼs error handling. This seems to be because `error_stack::Report<C>` doesnʼt impl `std::error::Error`, so it isnʼt compatible with any other error handling libraries.
>    It might be a bit inappropriate to use in a library, the same way anyhow is? Iʼll probably just try switching to it for now so I can `?` again.
>    
>    * It's also a little insidious in that the `Result<T, E>` return types in the documentation look normal, but then you realise that they're very much not!

I didn't know the fact that it didn't implement `std::error::Error` :-/. I agree that's unfortunate, even though I really like the interface. Guess I'll have to go nuts with `thiserror` :P.

> 4. The majority of the client crateʼs types donʼt seem to be exported publicly because all the modules are private - generate docs and look at all the broken links. For example: `EntityTopic`, `CommandTopic`, `ConnectError`, `CreateEntityError`, etc. Consider making the modules public?

I tend to by-default go private on everything, as that allows me to change things without it being a breaking change. I've not landed yet on what should and should not be public in the packages. But I agree, broken docs links should be removed, so i will not consider the API surface as satisfying before that's resolved.

> 5. Small nitpick, but `MqttQosLevel` probably doesnʼt need the `Mqtt` prefix

Agreed.

> 
> EDIT: some more thoughts...
> 
> * it'd be neat if `HassMqttClient` itself would impl `MqttClient` (or expose a `.client()` accessor), though obviously that's not a recommended interface to use normally.

The `MqttClient` interface is intended to be an implementation detail (and the details aren't finished yet). I'm not sure I want it to be public, but if you have a very good use-case I might change my mind.

> * the `mqtt` module types have a lot of `pub(crate)` fields and types going on that makes it impossible for a crate to actually impl `MqttClient`/`MqttProvider` externally.

It's very intentional that you can't implement the provider externally currently.  If that's to be changed, I would probably move the provider-traits to it's own crate - and a separate crate for `hass-mqtt-provider-paho` etc......
Thinking about it, I may just do that.

> * there could maybe be some QOL improvements if it's possible to impl `Into<Topic>` for `&CommandTopic`, `&StateTopic`, `&Arc<str>`, and `Arc<str>`. `&*` seems silly if it can be avoided.

Yes - there's definitely some improvements to be had there, given that a lot of the things takes `Into<X>` as arguments.

> * according to [hass docs](https://www.home-assistant.io/integrations/mqtt/#discovery-topic):
>   > Best practice for entities with a unique_id is to set <object_id> to unique_id and omit the <node_id>.
>   
>   
>   ... but there appears to be no way to omit the node_id from the generated discovery topic. I'm not against providing a node id anyway for the private prefix, but there could be a way to omit it from the config topic if one uses `client.entity_with_unique_id()` or something?

That's rather easy though - I can just do `${node_id}-${entity_id}` in the topic. I would like to "enfoce" a node-id, as (from experience) I've had the need to run 2 of certain integrations that made it ... tricky :P. 

> * on the same topic, `node_id` could be an `Option<String>` particularly when `unique_id` is provided instead (though I think technically both being `None` would still be a valid configuration?)

I much prefer enforcing some defaults - in general, I think the `hass-mqtt-client` should provide some opinions on things. If anyone want to overrule them, it's not that big of an abstraction over just using a MQTT library directly. The value-add in the client library is mostly the enforced defaults.

>   just in general I guess I'm looking for a bit more customization/control over the topic formatting. `entity.state_topic_with("state/blah")` and `command_topic_with("command/blah")`, for example. I also believe that `domain` doesn't need to be part of the private topic if a `unique_id` is used for the entity, since it must be globally unique?

See comment above. That being said, the formatting of the topic for the private topics could be changed.

> * I feel a little odd about the one-to-one relationship between a `CommandTopic` and subscriptions. I've set up a big [`futures::stream::select_all`](https://docs.rs/futures/0.3.25/futures/stream/fn.select_all.html) over 100 `CommandTopics` since I see no obvious way to use a wildcard, and it just feels awkward? If the router is so fancy, why can't `HassMqttClient impl Stream<Item = (TopicToken, Message)>` or something? stuck_out_tongue

If you take a look at how the subscription handling works, the idea is that you would do something similar to that. As for whether or not a single wildcard subscription should be made to the broker itself, that would be an internal implementation detail. Idea is that you'd do something like:

```rust
let mut subscriptions = Subscriptions::new();
subscriptions.push(light_command_subscription);
subscriptions.push(some_other_command_subscription);

while let Some((subscription, message)) = subscriptions.next() {
  // handle
}
```

though this is definitely not finalized. A lot of the details of the lower level changed when writing the client, and a lot of the details of the client is likely to change when I start writing the higher level library that I intend to use to actually implement lights etc. I plan to migrate https://github.com/YoloDev/enet-mqtt once I have something that seems like it could work, that should also work as an acid test.

>   I don't have any solid suggestions here though, haven't thought about it that much.
> * online/offline messages should also be something you can customize the payload and topic via the `Options` builder, but I'm assuming that's just an unfinished feature since there's an empty `availability` submodule

Same here with wanting to provide defaults. That being said, there's definitely some parts missing, as currently my example light does not provide availability info.

>   (also missing is a way to get the availability topic string from a `HassMqttClient`)

Precicely. Not just the channel, but the payloads as well.

**Comment by arcnmx** [comment]
> I didn't know the fact that it didn't implement `std::error::Error` :-/. I agree that's unfortunate, even though I really like the interface. Guess I'll have to go nuts with `thiserror` :P.

Yeah, it's a neat library but conversions to its error type seem to be one-way only - afaict you can't even get `C` back out of `Report<C>` if you tried (only a reference). For now I'm using it though, so!

> I tend to by-default go private on everything, as that allows me to change things without it being a breaking change. I've not landed yet on what should and should not be public in the packages. But I agree, broken docs links should be removed, so i will not consider the API surface as satisfying before that's resolved.

Sure, but broken links mean that you can get and use a `CommandTopic` as a return value from a function but can't actually name its type or store it in a struct, that's kind of a problem!

> > 5. Small nitpick, but `MqttQosLevel` probably doesnʼt need the `Mqtt` prefix
> 
> Agreed.

Other note: `hass_mqtt_types::qos::MqttQos` also exists so these can probably be unified, and the client crate could probably just re-export that?

> > * it'd be neat if `HassMqttClient` itself would impl `MqttClient` (or expose a `.client()` accessor), though obviously that's not a recommended interface to use normally.
> 
> The `MqttClient` interface is intended to be an implementation detail (and the details aren't finished yet). I'm not sure I want it to be public, but if you have a very good use-case I might change my mind.

A few reasons, none particularly compelling but not having the option can be limiting...

- The `MqttClient` traits to me are actually the most interesting part of the client library, as abstracting over paho/etc instead of using it directly is appealing to me, particularly because I did want to switch backends at some point.
- There can be other reasons to communicate over the same MQTT broker in an application that also needs to interface with home-assistant:
  - take for example [an application that triangulates position from ESPresense nodes](https://github.com/ESPresense/ad-espresense-ips): it needs to subscribe to custom MQTT topics in order to retrieve data from BLE beacon sensors, and then convey the accumulated data back to a home-assistant entity
    (note that particular implementation doesn't actually use the mqtt discovery protocol, but an alternative that performs the same function could and should!)
  - side-channel communication with other instances of the same application running on other machines/nodes
  - in general it seems silly to have to create two clients to the same broker if you use both the client crate and mqtt for any other communication? it just makes it much less modular and useful

> It's very intentional that you can't implement the provider externally currently. If that's to be changed, I would probably move the provider-traits to it's own crate - and a separate crate for `hass-mqtt-provider-paho` etc...... Thinking about it, I may just do that.

Not trying to make more work for you, but it could be a pretty nice general-purpose MQTT client crate if you did! It'd just be a shame if they remained private is all.

> That's rather easy though - I can just do `${node_id}-${entity_id}` in the topic. I would like to "enfoce" a node-id, as (from experience) I've had the need to run 2 of certain integrations that made it ... tricky :P.

Well, if an entity sets `unique_id` in the `Document` and serializes it, having distinct topics (allegedly) won't save you from home-assistant being angry at you for having two entities that share the same id. Physical devices might use a MAC address as their object id, because they only provide a single entity per domain type.

In some cases it could even be desired for multiple instances to provide the same entity as a sort of failover mechanism.

My main issue here is that in order for me to migrate my existing implementation to the client crate and remain compatible, I need to maintain the same config topic. Otherwise home-assistant will detect entities from the client crate as distinctly new ones and duplicate them, never quite forgetting about the prior implementation's entities without manual intervention (the mqtt integration is very temperamental about minor changes wrt IDs and topics).

> > * on the same topic, `node_id` could be an `Option<String>` particularly when `unique_id` is provided instead (though I think technically both being `None` would still be a valid configuration?)
> 
> I much prefer enforcing some defaults - in general, I think the `hass-mqtt-client` should provide some opinions on things. If anyone want to overrule them, it's not that big of an abstraction over just using a MQTT library directly. The value-add in the client library is mostly the enforced defaults.
> 
> That being said, the formatting of the topic for the private topics could be changed.

Defaults and conventions are well and good, but inflexibility and lacking modular design just makes it that much harder to migrate to while maintaining backward-compatibility.
I suppose the reason I argue is... home-assistant's MQTT protocol as a concept is just a way to enable auto-discovery on a broker for an MQTT service. It feels presumptuous to assume that such a service targets home-assistant first and foremost, with MQTT relegated to an implementation detail or simple transport. In my case, I'm aiming for providing a service that can be queried over MQTT, so the home-assistant entities/payloads are really just basic presets and pre-baked commands to provide a convenient UI over that.
Perhaps that does just make me not the audience of this crate, but I like to think that it could be helpful nonetheless.

It's also just personal preference, I feel that including the domain in the private prefix is overly verbose, redundant, and ugly. A `unique_id` is unique, it doesn't need disambiguation!
Though I'm cool with the node_id in the private prefix, but again, don't see why it can't be `None` if one feels strongly about the specific formatting.
(more thoughts below on this point)

**tldr;** I suppose I just feel that there should be the option to use either topic naming scheme, depending on whether one creates the entity with `client.entity(object_id)` or `client.entity_with_unique_id(uid)`. These are distinct concepts to the home-assistant mqtt integration, so why not provide the option?

> If you take a look at how the subscription handling works, the idea is that you would do something similar to that. As for whether or not a single wildcard subscription should be made to the broker itself, that would be an internal implementation detail. Idea is that you'd do something like:

A thought: if the client can guarantee that the `Arc` topic returned by the state is the same as the field in a `Message` from a topic `Stream`... You could just use that pointer as a lightweight token (comparing via [`Arc::ptr_eq`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.ptr_eq) and so on).

Perhaps a `Topic(Arc<str>)` newtype would be appropriate so that it could impl hash/eq/etc efficiently (without needing to touch the string data, as long as the [strings are interned](https://docs.rs/string-interner/latest/string_interner/)), as well as facilitate the various conversions to `hass_mqtt_types::Topic`. I'm not sure whether the type will need a different name (`TopicName`? `ClientTopic`? `MqttTopic`?) to disambiguate.

> > * online/offline messages should also be something you can customize the payload and topic via the `Options` builder, but I'm assuming that's just an unfinished feature since there's an empty `availability` submodule
> 
> Same here with wanting to provide defaults. That being said, there's definitely some parts missing, as currently my example light does not provide availability info.

\*shrugs\* my existing availability topic just doubles as a debug json attrs topic is all

> > (also missing is a way to get the availability topic string from a `HassMqttClient`)
> 
> Precicely. Not just the channel, but the payloads as well.

Sure, though since you're currently using the defaults, it's not *strictly* necessary since one can just throw the topic name into `availability_topic` and expect it to work as-is.

## state/command topics

`"state"` and `"command"` namespaces seem arbitrary and unnecessary. For the majority of entity types you have a maximum of two state and command topics, but usually you only have one - either a boring state_topic or, if you have a json attrs topic, chances are it doubles as a state topic with `value_template` extracting a value out of it. So I create `entity.state_topic("state")`, and I get a topic named `my/private/prefix/E34350705900/light/E34350705900/state/state`, excessive! I suppose my suggestions here are:
1. My expectation was that I'm naming the topic, so it's my responsibility to avoid using the same name for multiple topics (and that it would likely error if I tried). Thus, I didn't expect them to be arbitrarily "namespaced" under state or command.
2. If I wanted defaults (because I only have one such topic, like most users presumably will), `state_topic()` can just take no argument by default, and give me a topic named `prefix/id/state`
3. `with_state_topic(name)` seems appropriate in cases where one specifically wants to name it, or has to disambiguate multiple topics - in which case, it really doesn't need the namespace of `"state/{name}"`.
4. If one doesn't provide a private prefix, it might make sense to heed the mqtt integration's docs:
    > The `<node_id>` level can be used by clients to only subscribe to their own (command) topics by using one wildcard topic like `<discovery_prefix>/+/<node_id>/+/set`.
  
    So... if no private prefix exists, they seem to suggest using the entity discovery config topic (minus trailing `/config`) for your state/command topics, which might make some sense?
5. If this gets too complicated maybe it just means that the entities (and/or subtopics?) need an options builder too to configure formatting and things, ugh...

**Comment by Alxandr** [comment]
> Sure, but broken links mean that you can get and use a `CommandTopic` as a return value from a function but can't actually name its type or store it in a struct, that's kind of a problem!

Yes - I know that. It's why I said I don't consider the API to be satisfying before this is resolved. I want everything* that is reachable to be nameable, but I also want to be a bit careful when designing the public surface area.

> Other note: `hass_mqtt_types::qos::MqttQos` also exists so these can probably be unified, and the client crate could probably just re-export that?

Yes - will do that.

>   * side-channel communication with other instances of the same application running on other machines/nodes
>   * in general it seems silly to have to create two clients to the same broker if you use both the client crate and mqtt for any other communication? it just makes it much less modular and useful

These are good points. I don't really want to add generics to the client though - so I think I'll allow you to create a `HassMqttClient` from `impl MqttClient` - that should be a good compromise. Before I do so though, I'll need to have a think about how to deal with online/offline messages.

> Not trying to make more work for you, but it could be a pretty nice general-purpose MQTT client crate if you did! It'd just be a shame if they remained private is all.

Extracting the abstraction I've made so far to it's own crate should be fairly easy. I think I'll just do that.

> > That's rather easy though - I can just do `${node_id}-${entity_id}` in the topic. I would like to "enfoce" a node-id, as (from experience) I've had the need to run 2 of certain integrations that made it ... tricky :P.
> 
> Well, if an entity sets `unique_id` in the `Document` and serializes it, having distinct topics (allegedly) won't save you from home-assistant being angry at you for having two entities that share the same id. Physical devices might use a MAC address as their object id, because they only provide a single entity per domain type.
> 
> In some cases it could even be desired for multiple instances to provide the same entity as a sort of failover mechanism.
> 
> My main issue here is that in order for me to migrate my existing implementation to the client crate and remain compatible, I need to maintain the same config topic. Otherwise home-assistant will detect entities from the client crate as distinctly new ones and duplicate them, never quite forgetting about the prior implementation's entities without manual intervention (the mqtt integration is very temperamental about minor changes wrt IDs and topics).

This is also a rather good point... Maybe I could create a function that allows you to manually specify the topic? `entity.state_topic("foo").override_topic("foo/bar/baz")`? And if you do, you're completely on your own wrt not stepping on your own shoes. It would make using it *slightly* more annoying than the default safe option. Probably also document that it's suggested not to use, but only there if you know what you're doing and have a good reason?

**Comment by Alxandr** [comment]
> `"state"` and `"command"` namespaces seem arbitrary and unnecessary. For the majority of entity types you have a maximum of two state and command topics, but usually you only have one - either a boring state_topic or, if you have a json attrs topic, chances are it doubles as a state topic with `value_template` extracting a value out of it. So I create `entity.state_topic("state")`, and I get a topic named `my/private/prefix/E34350705900/light/E34350705900/state/state`, excessive! I suppose my suggestions here are:

I completely agree that it's excessive (and I also kind of hate it). However I kind of wanted to take a rustfmt/prettier stand on it where it's not supposed to be "formatted" by the developer. There is one "canonical" format that's "enforce" on you - and though it may not be the most pretty, it's easy to validate. That being said - I've been convinced to allow you to set it yourself.

>    So... if no private prefix exists, they seem to suggest using the entity discovery config topic (minus trailing `/config`) for your state/command topics, which might make some sense?

All plugins I've ever seen uses it's own private prefix. I think I'm just going to stick to that.

> 5. If this gets too complicated maybe it just means that the entities (and/or subtopics?) need an options builder too to configure formatting and things, ugh...

It's not like we have no experience building builders at this point, no? xD

Also - I probably missed something, but this PR can soon be published as a novel xD - so do forgive me.

File: crates/mqtt-types/src/validation.rs [hunk 1]
```diff
@@ -1,17 +1,65 @@
-use error_stack::Context;
 use semval::{Invalidity, Validate};
-use std::fmt;
+use std::{backtrace::Backtrace, fmt};
```
[the remark this answers is not in the document]
Comment (Alxandr) [reply crates/mqtt-types/src/validation.rs:2]: Given that it's in stable already, I think we can just feature gate it? That way you can turn of backtrace-support if you don't want it even on latest version.

**Comment by Alxandr** [comment]
You wanted crates @arcnmx  :D

**Comment by Alxandr** [comment]
```
❯ cargo deadlinks
 Documenting hass-mqtt-provider-paho v0.0.0 (/home/alxandr/hub/hass-rs/crates/mqtt-provider-paho)
 Documenting hass-mqtt-provider v0.0.0 (/home/alxandr/hub/hass-rs/crates/mqtt-provider)
 Documenting hass-dyn-error v0.0.0 (/home/alxandr/hub/hass-rs/crates/dyn-error)
 Documenting hass-provide-any-probe v0.0.0 (/home/alxandr/hub/hass-rs/build/provide-any-probe)
 Documenting hass-mqtt-client v0.0.0 (/home/alxandr/hub/hass-rs/crates/mqtt-client)
 Documenting hass-mqtt-types v0.2.0 (/home/alxandr/hub/hass-rs/crates/mqtt-types)
 Documenting mqtt-light v0.0.0 (/home/alxandr/hub/hass-rs/examples/mqtt-light)
 Documenting hass-mqtt-macros v0.0.0 (/home/alxandr/hub/hass-rs/crates/mqtt-macros)
    Finished dev [unoptimized + debuginfo] target(s) in 1.96s
Found invalid urls in entity/enum.ColorMode.html:
	Fragment #tymethod.deserialize at entity/enum.ColorMode.html does not exist!
	Fragment #tymethod.serialize at entity/enum.ColorMode.html does not exist!
	Fragment #tymethod.enum_from_u32 at entity/enum.ColorMode.html does not exist!
	Fragment #tymethod.enum_into_u32 at entity/enum.ColorMode.html does not exist!
```

Not sure these last broken links are something I can do anything about. They are from external traits as far as I can tell.

**Comment by Alxandr** [comment]
@arcnmx this PR has gotten enough messages that github has started hiding the middle :stuck_out_tongue: - that being said I don't remember anymore all of the things I was going to do, so I added a few checkboxes to the first message. The only point I currently have left is allowing you to control the MQTT topics *if you so desire*. Not landed entirely on how yet, but I'm about to start looking at that. Once that's done I'm thinking I'll merge this, and you can open issues for things that are missing, cause it's so much easier to keep track off than a long list of text.

**Pull request merged**